### PR TITLE
fixes #18322 - provide a way to reset certs to self-signed

### PR DESCRIPTION
**Pull request by stbenjam**
This PR adds an option called "--certs-reset" that will remove
any settings related to custom certs and then cause a new
self-signed CA to be created.

### CODE REVIEW

**Comment by stbenjam** [comment]
This works.   If you want to test, install a server with custom certs.  You can use https://github.com/Katello/forklift/pull/388 to do it.

```
ansible-playbook -l <box> playbooks/custom_certificates.yml
```

And then run the installer:

```
foreman-installer -v --scenario katello --certs-server-cert /root/custom_certificates/certs/rhel7.example.com.crt --certs-server-cert-req /root/custom_certificates/requests/rhel7.example.com.csr --certs-server-key /root/custom_certificates/private/rhel7.example.com.key  --certs-server-ca-cert /root/custom_certificates/certs/custom_ca.crt
```

Do a hammer ping, visit the web ui, publish a CV, everything's a-OK. Then reset certs to self signed:

```
foreman-installer --certs-reset
```

Then do a hammer ping, look at the web ui and notice it's using self-signed certs, and that everything is a-OK too.

**Comment by chris1984** [comment]
@stbenjam will test this and report back

**Comment by stbenjam** [comment]
@chris1984 Thanks, also this is the box I was using.  The main thing to do is make sure the installer is skipped (foreman_installer_skip_installer) so you can install with custom certs.

```
certs2:
  box: centos7
  memory: 6144
  ansible:
    playbook: 'playbooks/katello.yml'
    group: 'server'
    variables:
        katello_version: nightly
        foreman_version: nightly
        foreman_installer_skip_installer: True
```

**Comment by chris1984** [comment]
Box Setup:

Thursday 16 February 2017  18:21:06 -0500 (0:00:00.016)       0:06:55.815 ***** 
=============================================================================== 
foreman_installer : Install additional packages ----------------------- 272.33s
foreman_installer : Update packages ------------------------------------ 86.18s
foreman_installer : Install foreman-installer -------------------------- 44.05s
foreman_installer : Install foreman-release-scl ------------------------- 5.39s
epel_repositories : Setup Epel Repository ------------------------------- 1.90s
katello_repositories : Setup Katello Repository ------------------------- 1.68s
puppet_repositories : Setup Puppet 3 Repository ------------------------- 1.08s
foreman_repositories : Setup Foreman Repository ------------------------- 1.08s
setup ------------------------------------------------------------------- 0.87s
selinux : Ensure libselinux-python is installed ------------------------- 0.46s
selinux : Set selinux state --------------------------------------------- 0.24s
etc_hosts : Build hosts file -------------------------------------------- 0.22s
katello_repositories : Set up qpid Copr repository ---------------------- 0.02s
foreman_installer : install module PR into installer directory ---------- 0.02s
foreman_repositories : Setup Foreman Repository ------------------------- 0.02s
foreman_installer : install puppetlabs-strings -------------------------- 0.02s
foreman_installer : Run installer --------------------------------------- 0.02s
foreman_installer : Update foreman-release-scl -------------------------- 0.02s
puppet_repositories : Setup Puppet 4 Repository ------------------------- 0.02s
foreman_installer : Update packages ------------------------------------- 0.02s

Running certs playbook:

chrobert@vault111:~/Dropbox/git/forklift$ ansible-playbook -l certs2 playbooks/custom_certificates.yml 

PLAY RECAP *********************************************************************
certs2                     : ok=12   changed=10   unreachable=0    failed=0   

Thursday 16 February 2017  18:30:23 -0500 (0:00:00.173)       0:00:06.002 ***** 
=============================================================================== 
setup ------------------------------------------------------------------- 3.18s
custom_certificates : Deploy configuration file ------------------------- 0.46s
custom_certificates : Install openssl ----------------------------------- 0.41s
custom_certificates : Ensure serial starting number --------------------- 0.34s
custom_certificates : Creating CA certificate and key ------------------- 0.33s
custom_certificates : Create certs directory ---------------------------- 0.23s
custom_certificates : Create server key --------------------------------- 0.18s
custom_certificates : Create server certificate ------------------------- 0.17s
custom_certificates : Create server signing request --------------------- 0.17s
custom_certificates : Create requests directory ------------------------- 0.17s
custom_certificates : Create keys directory ----------------------------- 0.17s
custom_certificates : Create index file --------------------------------- 0.16s

Running installer:

foreman-installer -v --scenario katello --certs-server-cert /root/custom_certificates/certs/certs2.example.com.crt --certs-server-cert-req /root/custom_certificates/requests/certs2.example.com.csr --certs-server-key /root/custom_certificates/private/certs2.example.com.key  --certs-server-ca-cert /root/custom_certificates/certs/custom_ca.crt --disable-system-checks

Katello installer with custom ssl certs:

[ INFO 2017-02-16 23:41:17 verbose] Executing hooks in group post
  Success!
  * Katello is running at https://certs2.example.com
      Initial credentials are admin / dNsoAJ5CJz4qVU7o
  * To install an additional Foreman proxy on separate machine continue by running:

      foreman-proxy-certs-generate --foreman-proxy-fqdn "$FOREMAN_PROXY" --certs-tar "~/$FOREMAN_PROXY-certs.tar"

  The full log is at /var/log/foreman-installer/katello.log
[ INFO 2017-02-16 23:41:18 verbose] pulp.conf is already present, skipping
[ INFO 2017-02-16 23:41:18 verbose] All hooks in group post finished

[root@certs2 pre]# hammer ping
candlepin:      
    Status:          ok
    Server Response: Duration: 22ms
candlepin_auth: 
    Status:          ok
    Server Response: Duration: 22ms
pulp:           
    Status:          ok
    Server Response: Duration: 61ms
foreman_tasks:  
    Status:          ok
    Server Response: Duration: 36ms

Using custom cert:

https://www.dropbox.com/s/34531crdh7xdcr0/custom_cert.png?dl=0

content view created ok:

https://nimbus.everhelper.me/client/notes/share/771962/gy29o7mkciogd0ena32u

reseting certs now:

foreman-installer --certs-reset -v

[ INFO 2017-02-17 00:10:21 verbose] Puppet has finished, bye!
[ INFO 2017-02-17 00:10:21 verbose] Executing hooks in group post
  Success!
  * Katello is running at https://certs2.example.com
  * To install an additional Foreman proxy on separate machine continue by running:

      foreman-proxy-certs-generate --foreman-proxy-fqdn "$FOREMAN_PROXY" --certs-tar "~/$FOREMAN_PROXY-certs.tar"

  The full log is at /var/log/foreman-installer/katello.log
[ INFO 2017-02-17 00:10:21 verbose] pulp.conf is already present, skipping
[ INFO 2017-02-17 00:10:21 verbose] All hooks in group post finished

[root@certs2 pre]# hammer ping
candlepin:      
    Status:          ok
    Server Response: Duration: 21ms
candlepin_auth: 
    Status:          ok
    Server Response: Duration: 22ms
pulp:           
    Status:          ok
    Server Response: Duration: 66ms
foreman_tasks:  
    Status:          ok
    Server Response: Duration: 35ms


Using new self-signed cert: 

https://www.dropbox.com/home?preview=new_cert.png

New content view showing its working:

https://www.dropbox.com/s/zi5vr757c1ss657/view_after_newcerts.png?dl=0

File: hooks/boot/20-certs_update.rb [hunk 1]
```diff
@@ -18,6 +18,14 @@
   :default => false
 )
 app_option(
+  '--certs-reset',
+  :flag,
+  "This option will reset any custom certificates and use the self-signed CA " \
+  "instead. Note that any clients will need to be updated with the latest " \
+  "katello-ca-consumer RPM.",
```
Comment (ehelms) [inline hooks/boot/20-certs_update.rb]: 1) Does this cause a need to regen and re-do Proxies as well?
 2) If I reset my server back to the default CA, can I leave my Proxies on the custom certs?

Comment (stbenjam) [inline hooks/boot/20-certs_update.rb]: Pfft, you and your questions!  I think (1) and (2) should be OK, but l'll actually test with a proxy to make sure.

Comment (stbenjam) [inline hooks/boot/20-certs_update.rb]: 1. Yes:

After doing a `--certs-reset` on the Katello, you need to generate a new certs tarball and run the installer on the proxy, too.  After doing that the external proxy will work again.

2. No, you need to do the steps above

Comment (chris1984) [inline hooks/boot/20-certs_update.rb]: Is it possible to add that info wrt smart-proxy tar file into the info above, so the user knows he has to update his external proxies?

Comment (stbenjam) [inline hooks/boot/20-certs_update.rb]: I already added it to the --certs-reset help, and I'll update the documentation on the web site

**Comment by stbenjam** [comment]
Thanks @ehelms, @chris1984!

**Pull request merged**